File: packages/reactive_ble_mobile/android/src/main/kotlin/com/signify/hue/flutterreactiveble/utils/BleLogger.kt
package com.signify.hue.flutterreactiveble.utils

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

internal object BleLogger {
    @Volatile
    var eventSink: EventChannel.EventSink? = null

    private val mainHandler = Handler(Looper.getMainLooper())

    fun info(
        source: String,
        message: String,
    ) = emit(source, "Info", message)

    fun warning(
        source: String,
        message: String,
    ) = emit(source, "Warning", message)

    fun error(
        source: String,
        message: String,
    ) = emit(source, "Error", message)

    private fun emit(
        source: String,
        level: String,
        message: String,
    ) {
        val sink = eventSink ?: return
        val payload =
            mapOf(
                "Level" to level,
                "Source" to source,
                "Message" to message,
            )
        if (Looper.myLooper() == Looper.getMainLooper()) {
            sink.success(payload)
        } else {
            mainHandler.post { eventSink?.success(payload) }
        }
    }
}

File: packages/reactive_ble_mobile/darwin/reactive_ble_mobile/Sources/reactive_ble_mobile/Plugin/BleLogger.swift
#if os(iOS)
  import Flutter
#elseif os(macOS)
  import FlutterMacOS
#endif
import Foundation

enum BleLogger {
    static var eventSink: EventSink?

    static func info(_ source: String, _ message: String) {
        emit(source: source, level: "Info", message: message)
    }

    static func warning(_ source: String, _ message: String) {
        emit(source: source, level: "Warning", message: message)
    }

    static func error(_ source: String, _ message: String) {
        emit(source: source, level: "Error", message: message)
    }

    private static func emit(source: String, level: String, message: String) {
        eventSink?.addLog([
            "Level": level,
            "Source": source,
            "Message": message,
        ])
    }
}
